DistinctOperation now returns a batch Cursor

rozza opened the pull request:
JAVA-1624

jyemin commented on bson/src/main/org/bson/codecs/BsonArrayCodec.java:
```diff
@@ -21,26 +21,41 @@
+    /**
+     * Creates a new instance with a default {@link org.bson.codecs.configuration.RootCodecRegistry}
+     */
+    public BsonArrayCodec() {
```
This constructor appears to be unused.  Why did you go to the trouble of creating it?

rozza replied:
Follows `BsonDocumentCodec` strictly speaking I'm not sure we need to pass an alternative codecRegistry in.

jyemin commented on driver-core/src/main/com/mongodb/operation/BsonArrayWrapper.java:
```diff
@@ -0,0 +1,233 @@
+import static org.bson.assertions.Assertions.notNull;
+
+
+class BsonArrayWrapper<T> extends BsonArray {
```
Most of this class appears to be unused and untested.  In particular, getUnwrapped is never called.

jyemin commented on driver-core/src/main/com/mongodb/operation/BsonArrayWriter.java:
```diff
@@ -0,0 +1,241 @@
+import static org.bson.BsonContextType.SCOPE_DOCUMENT;
+
+
+class BsonArrayWriter extends AbstractBsonWriter {
```
Called by unwrapped above, but never tested.

jyemin commented:
Looks good.  As for the un-exercised code, since it's all in non-public classes, another option (besides just writing the tests) is to just remove it and replace with throw new UnsupportedOperationException().  Otherwise,

jyemin commented:
Proposing a simplification of BsonValueCodec, as I found its dependency on BsonDocumentCodec confusing.

rozza commented:
Looks better thanks :+1: 

I've updated and `BsonArrayWrapper` is not much of a wrapper but we can always implement the missing features should the need ever arise.  I changed the `BsonWrapperHelper.toList` method signature to try and make it more palatable.

jyemin commented:
lgtm

rozza commented:
rebased an merged
